tools/linter/adapters/perf_linter.py

```diff
@@ -0,0 +1,248 @@
+from __future__ import annotations
+
+import enum
+import sys
+import token
+from pathlib import Path
+from typing import TYPE_CHECKING
+
+
+_PARENT = Path(__file__).parent.absolute()
+_PATH = [Path(p).absolute() for p in sys.path]
+
+if TYPE_CHECKING or _PARENT not in _PATH:
+    from . import _linter
+else:
+    import _linter
+
+if TYPE_CHECKING:
+    from collections.abc import Iterator
+    from tokenize import TokenInfo
+
+
+DESCRIPTION = "`perf_linter` finds common Python patterns which are slow and suggests faster alternatives."
+EPILOG = """
+You can use this either via `lintrunner` or directly, via:
+
+    python tools/linter/adapters/perf_linter.py --fix [... python files ...]
+
+---
+
+To omit a line of Python code from `perf_linter` checking, append a comment:
+
+    s = ",".join(x for x in range(3))  # noqa: perf_linter
+
+---
+
+perf_linter makes the following suggestions:
+
+1) Generators to list comprehensions:
+
+    >>> timeit.timeit("','.join(str(x) for x in range(10))")
+    0.978244581958279
+    >>> timeit.timeit("','.join([str(x) for x in range(10)])")
+    0.695843324996531
+
+2) List constructors:
+
+    >>> timeit.timeit("list(range(2))")
+    0.10411133896559477
+    >>> timeit.timeit("[*range(2)]")
+    0.08048266009427607
+
+3) Empty list/dict constructors:
+
+    >>> timeit.timeit('list()', number=10000000)
+    0.38338504498824477
+    >>> timeit.timeit('[]', number=10000000)
+    0.18809661595150828
+    >>> timeit.timeit('dict()', number=10000000)
+    0.3867137918714434
+    >>> timeit.timeit('{}', number=10000000)
+    0.17520155780948699
+
+"""
+ERROR1 = "Generators are slow! Use list comprehensions instead."
+ERROR2 = "list(x) is slow! Use [*x] instead."
+ERROR3 = "list()/dict() is slow! Use []/{} instead."
```

Comment:
Do not reinvent the wheel here: C408 already does this. It's just disabled because "we like the old dict kwarg syntax` even though it's slower, less efficient etc.